### PR TITLE
Implement some additional features and merge some existing PR

### PR DESCRIPTION
**Pull request by dukesun99**
Changes made:
- Support number type node and edge caption
- Implement the to-do image for nodes feature (As stated in readme)
- Ability to update visualization with Cypher (Merged from #104, reviewed and tested)
- Add support for connecting to Neo4j 4 databases (Merged from #110, reviewed but not really know how to test the new feature, @zakjan please help!)
- Configurable font size (As requested by #108)
- Modify index.d.ts file (#102)

@thebestnom @johnymontana Please help to review the code, thanks!

### CODE REVIEW

**Comment by dukesun99** [comment]
I merged #104 to my branch also. Hope this can reduce some of your review and merge work.

**Comment by dukesun99** [comment]
I am done with my current work. I feel it would be good to support `image cypher` as it is also requested by #80.   
I feel it is also good to rethink if we should expose some of the vis.js config API to users. (Font as an example)

**Comment by zakjan** [comment]
@dukesun99 Re my PR, it's simple to test! Create a new database with `CREATE DATABASE <name>` query, switch to it by `:use <name>`, create some data in it. Configure Neo4j driver to use the new database by the new `server_database` neovis param. You should see the data from the new database, instead of the default database.

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,4 @@ node_modules
 docs/build
 /lib
 .npmrc
+examples/neovis.html
```
Comment (thebestnom) [inline .gitignore:14]: why do you ignore this file?

File: README.md [hunk 2]
```diff
@@ -309,6 +317,11 @@ import { NEOVIS_DEFAULT_CONFIG } from 'neovis.js';
     "caption": "name",
     "size": "pagerank",
     "community": "community",
+    //"image": 'https://visjs.org/images/visjs_logo.png',
```
Comment (thebestnom) [inline README.md:320]: Why is that commented?

File: package.json [hunk 3]
```diff
@@ -59,7 +59,7 @@
   },
   "dependencies": {
     "@babel/runtime-corejs3": "^7.8.4",
-    "neo4j-driver": "^1.7.5",
+    "neo4j-driver": "^4.1.0",
```
Comment (thebestnom) [inline package.json:62]: I think it's time to move that to peerDependencies, but maybe I'll do that in a different PR

File: src/neovis.js [hunk 4]
```diff
@@ -158,6 +162,8 @@ export default class NeoVis {
 		// node caption
 		if (typeof captionKey === 'function') {
 			node.label = captionKey(neo4jNode);
+		} else if (captionKey && (typeof neo4jNode.properties[captionKey] === 'number' || Neo4j.isInt(neo4jNode.properties[captionKey]))) {
```
Comment (thebestnom) [inline src/neovis.js:165]: why `Neo4j.isInt` is not enough?

File: src/neovis.js [hunk 5]
```diff
@@ -253,16 +276,16 @@ export default class NeoVis {
...
 
-	render() {
+	render(query) {
 
 		// connect to Neo4j instance
 		// run query
 		let recordCount = 0;
-
-		let session = this._driver.session();
+		const _query = query ? query : this._query;
```
Comment (thebestnom) [inline src/neovis.js:284]: I think as you should do `query || this._query` to make that more consistent with the rest of the code

File: src/neovis.js (hunk 4 above)
Comment (dukesun99) [reply src/neovis.js:165]: isInt only checks neo4j int type if I am not wrong. To handle float we need to do another number check.

File: src/neovis.js (hunk 5 above)
Comment (dukesun99) [reply src/neovis.js:284]: Ok will fix

File: README.md (hunk 2 above)
Comment (dukesun99) [reply README.md:320]: I feel some people would think `image` is a mandatory field if we do not comment it out.

File: .gitignore (hunk 1 above)
Comment (dukesun99) [reply .gitignore:14]: That's my mistake. Will fix.

**Comment by dukesun99** [comment]
Test added for the new parts I added. I think we should include how to run the test in the readme as I even not realized we have some tests...

**Comment by dukesun99** [comment]
@zakjan Thank you. I tested but I got `Neo4jError: Queries cannot be run directly on a session with an open transaction; either run from within the transaction or use a different session.` if I build with `neo4j-driver` version `4.1.0`.   
May I know do you have this issue on your side or I missed something?

File: src/neovis.js (hunk 4 above)
Comment (thebestnom) [reply src/neovis.js:165]: Oh, I toght there was a `&&` there, my bad

**Comment by zakjan** [comment]
@dukesun99 It works for me in the included examples/example-labels.html. Could you add steps to replicate your issue? Is it possible that some other code is incompatible with the latest driver?

**Comment by thebestnom** [comment]
Just to make sure, the 4.1 driver works only for 4.1+ neo4j, what version of neo4j do you use?

**Comment by thebestnom** [comment]
Also, this PR is pretty big as is, please move the event to a separate PR

**Comment by zakjan** [comment]
Well, I wanted to make the driver dependency >=4.0.0, but it always installs the latest version anyway. It might make sense to move the driver to peerDependencies for better control.

**Comment by dukesun99** [comment]
@zakjan @thebestnom I guess would be the version problem? I am running 4.0.4.

**Comment by dukesun99** [comment]
> Also, this PR is pretty big as is, please move the event to a separate PR

Ok will do.

**Comment by thebestnom** [comment]
> @zakjan  Well, I wanted to make the driver dependency >=4.0.0, but it always installs the latest version anyway. It might make sense to move the driver to peerDependencies for better control.

Ill make it peer dependency in a different PR

**Comment by dukesun99** [comment]
I tried to build with 4.0.4 driver and I still get the same error. I cloned and built the repo from @zakjan , then I run my visualization on that repo (not using the new feature) and I got that error.

**Comment by thebestnom** [comment]
@dukesun99 give me a moment and ill test that

**Comment by thebestnom** [comment]
well, I trying using neo4j sandbox but the only version that I get there is 3.5.0, and there everything work as intended, Ill try later on newer Neo4j

**Comment by thebestnom** [comment]
sorry, I can't test this... I can't install neo4j on this PC, and I can't find any good hosting file that I can host neo4j 4.0.4 on

**Comment by dukesun99** [comment]
@thebestnom @zakjan I finally located the bug. It is because of the sizeCypher. That's where the concurrent access of the session happens. I will push a fix later.

**Comment by dukesun99** [comment]
@thebestnom I checked the history, you changed sizeCypher query from every time a new session to resuing the old session. I tried to change back to new session every node and the error is fixed. I think maybe we can refactor the code to do all sizeCypher queries in one transaction commit if you do have an efficiency concern.

**Comment by thebestnom** [comment]
the problem is that each session opens a new websocket, and browsers (specificly tested on chrome) gives a hard limit on maximum opened sockets (see https://github.com/neo4j-contrib/neovis.js/issues/85) and that worked until now, and it would be great if it could happen on the same websocket

**Comment by dukesun99** [comment]
@thebestnom Then I guess we need to do a bigger refactor and make use of transactions to fix that.

**Comment by thebestnom** [comment]
how much time would it take to change it from session to transaction?

**Comment by dukesun99** [comment]
@thebestnom I read through the driver document and I found an easier workaround for #85. The sessions are taken from a connection pool, which size is unlimited by default. We can limit that by providing the max size in the driver config object. I will push a fix.

**Comment by thebestnom** [comment]
@dukesun99 sounds great! thank you!

**Comment by dukesun99** [comment]
@thebestnom Ok I've fixed that. I set the max session number to 100. I tried 200 but I got some strange behavior. 100 is working fine.

**Comment by thebestnom** [comment]
Thanks! what is the reason you still used transaction anyway?

**Comment by dukesun99** [comment]
> Thanks! what is the reason you still used transaction anyway (instead of run)?

They say `Auto-commit transactions` is not recommended in production (for not having retry). See [here](https://neo4j.com/docs/driver-manual/current/session-api/asynchronous/#driver-async-autocommit-transactions)

**Comment by dukesun99** [comment]
Ok I finally added more descriptions in the readme. I think it's ready now. Please help to review and I will work on events then. @thebestnom

**Comment by thebestnom** [comment]
yes, it's looks ok as I said earlier, already contacted @johnymontana as I don't have merge permission here

as for event, there is already a way to implement it, so Im not sure what @johnymontana think of this

**Comment by dukesun99** [comment]
@thebestnom ok thank you. I need to use event in my own project so I just added it anyway, using the current event structure. The structure is there but now it's incomplete.

**Review by jexp — APPROVED** [review]
Looks great, thanks a lot!

**Comment by jexp** [comment]
Thank you all @dukesun99 @zakjan @thebestnom for your contributions!!

**Pull request merged**